pkg/cmd/render/render.go

```diff
@@ -109,98 +134,218 @@ func runRenderCmd(cmd *cobra.Command, args []string) {
+	}
+	if !isValidMode(installConfigMode) {
+		return fmt.Errorf("invalid mode defined: %s", installConfigMode)
+	}
+
+	installConfigModeManual := installConfigMode == string(operatorv1.CloudCredentialsModeManual)
+
+	// Make sure deprecated configmap setting doesn't conflict with install-config contents.
+	// Otherwise there could be a race between the configmap and the operator config
+	// resources landing causing unpredictable runtime behavior of the CCO pod.
+	if operatorDisabledViaConfigmap && installConfigMode != "" {
+		if operatorDisabledViaConfigmap != installConfigModeManual {
+			return fmt.Errorf("config map asking for CCO to be disabled, and install-config asking for %s mode", installConfigMode)
+		}
+	}
+
+	operatorDisabled := operatorDisabledViaConfigmap || installConfigMode == "manual"
 
 	log.Infof("Rendering files to %s", renderOpts.destinationDir)
 
 	ccoRenderDir := renderOpts.destinationDir
 
-	// render manifests
+	// render manifests from bindata
 	if err := os.MkdirAll(filepath.Join(ccoRenderDir, manifestsDir), 0775); err != nil {
-		log.WithError(err).Fatal("error creating manifests directory")
+		return errors.Wrap(err, "error creating manifests directory")
 	}
 
 	for _, assetName := range renderAssets {
 		asset, err := assets.Asset(assetName)
 		if err != nil {
-			log.WithError(err).Fatal("failed to read static asset")
+			return errors.Wrap(err, "failed to read static asset")
 		}
 
-		assetRenderPath := filepath.Join(ccoRenderDir, "manifests", "cco-"+filepath.Base(assetName))
-		log.Infof("Writing file: %s", assetRenderPath)
-		err = ioutil.WriteFile(assetRenderPath, asset, 0644)
-		if err != nil {
-			log.WithError(err).Fatal("failed to write file")
+		assetRenderPath := filepath.Join(ccoRenderDir, manifestsDir, "cco-"+filepath.Base(assetName))
+		if err := writeFile(assetRenderPath, asset); err != nil {
+			return err
 		}
 	}
 
+	// render operator config
+	var operatorConfig bytes.Buffer
+	templateVars := operatorTemplateVars{
+		ForceCredentialsMode: installConfigMode,
+	}
+
+	// if either the legacy config map or the install-config have CCO as disabled/manual,
+	// then we must set the config appropriately.
+	if operatorDisabled {
+		templateVars.ForceCredentialsMode = string(operatorv1.CloudCredentialsModeManual)
+	}
+
+	if err := operatorConfigTemplate.Execute(&operatorConfig, templateVars); err != nil {
+		return errors.Wrap(err, "failed to execute operator config template")
+	}
+	assetRenderPath := filepath.Join(ccoRenderDir, manifestsDir, operatorConfigFilename)
+	if err := writeFile(assetRenderPath, operatorConfig.Bytes()); err != nil {
+		return err
+	}
+
 	// need at least the empty dir so the installer bootkube.sh script works as expected
 	if err := os.Mkdir(filepath.Join(ccoRenderDir, bootstrapManifestsDir), 0775); err != nil {
-		log.WithError(err).Fatal("error creating bootstrap-manifests directory")
+		errors.Wrap(err, "error creating bootstrap-manifests directory")
 	}
 	if !operatorDisabled {
```

Comment:
Outside of the scope of this work, but do we care if the static pod is run when disabled? Since we will now be waiting for the config CR to exist, there is no longer the possibility of the static pod winning a race and incorrectly doing work.

Reply:
This is a good point. I'll write up a card to explore retiring this conditional launch logic.